alibi_detect/utils/schemas.py

```diff
@@ -0,0 +1,431 @@
+# mypy: ignore-errors
+# TODO - conditional checks depending on backend
+# TODO - consider validating output of get_config calls
+
+import numpy as np
+from pydantic import BaseModel
+from typing import Optional, Union, Dict, List, Callable
+from alibi_detect.utils._types import Literal
+from alibi_detect.version import __version__, __config_spec__
+from alibi_detect.cd.tensorflow import HiddenOutput, UAE
+from alibi_detect.models.tensorflow import TransformerEmbedding
+from transformers import PreTrainedTokenizerBase
+from alibi_detect.utils.frameworks import has_pytorch, has_tensorflow  # , has_sklearn
+
+
+SUPPORTED_MODELS = Union['UAE', 'HiddenOutput', 'tf.keras.Model']
+SupportedModels = [UAE, HiddenOutput]
+if has_tensorflow:
+    import tensorflow as tf
+    SupportedModels.append(tf.keras.Model)
+if has_pytorch:
+    import torch
+#    SupportedModels.append()  # TODO
+# if has_sklearn:
+#    SupportedModels.append()  # TODO
+SupportedModels = tuple(SupportedModels)
+
+
+# Custom BaseModel so that we can set default config
+class CustomBaseModel(BaseModel):
+    """
+    Base pydantic model schema. Set our pydantic default settings in here.
+    """
+    class Config:
+        arbitrary_types_allowed = True  # since we have np.ndarray's etc
+        extra = 'forbid'  # Forbid extra fields so that we catch misspelled fields
+
+
+class DetectorConfig(CustomBaseModel):
+    """
+    Base detector config schema. Only fields universal across all detectors should be defined in here.
+    """
+    name: str
+    version: str = __version__
+    config_spec: str = __config_spec__
+    backend: Literal['tensorflow', 'pytorch', 'sklearn'] = 'tensorflow'
+    # Note: Although not all detectors have a backend, we define in base class as `backend` also determines
+    #  whether tf or torch models used for preprocess_fn.
+
+
+class ModelConfig(CustomBaseModel):
+    """
+    Unresolved schema for (ML) models. Note that the model "backend" e.g. 'tensorflow', 'pytorch', 'sklearn', is set
+    by backend in DetectorConfig.
+    """
+    type: Literal['custom', 'HiddenOutput', 'UAE'] = 'custom'
+    src: str
+    custom_obj: Optional[dict] = None
+
+
+class EmbeddingConfig(CustomBaseModel):
+    """
+    Unresolved schema for text embeddings.
+    """
+    type: Literal['pooler_output', 'last_hidden_state',
+                  'hidden_state', 'hidden_state_cls']  # See alibi_detect.models.tensorflow.embedding
+    layers: Optional[List[int]] = None  # TODO - add check conditional on embedding type (see docstring in above)
```

Comment:
Not sure what the TODO/docstring refers to here.

Reply:
No idea what docstring that is referring to. WIll remove that bit of comment. The TODO about conditional checks is related to some of the discussion below, so will defer to then.

Reply:
Ah sorry spoke to soon, the TODO is referring to the docstring for [alibi_detect.models.tensorflow.embedding](https://docs.seldon.io/projects/alibi-detect/en/latest/api/alibi_detect.models.tensorflow.embedding.html). The expected type for the `layers` kwarg is conditional on the embedding `type`. Not 100% sure if we can even do conditional things like this in pydantic models, but added a TODO as something to look into.